From b231ad46b31af93a9aad9573bef673a1f7a261c1 Mon Sep 17 00:00:00 2001
From: Hariprasad Kelam <hkelam@marvell.com>
Date: Sun, 23 May 2021 14:29:43 +0530
Subject: [PATCH 1542/1921] octeontx2-af: derive correct DMAC filter index

AF driver assigns number of DMAC filters per lmac based
on simple formula max dmac filters/number of lmacs.
Number of DMAC filters per lmac vary based on number of
enabled lmacs.

current implementation is such that DMAC index associated
with lmac is derived based on lmac_id. It results wrong index
in case of non contiguous lmac. This patch corrects the
index calculation by using sequence id of lmac instead of
lmac_id.

Change-Id: I8c3cca0016b5720b032f6f15f8d08429908eee29
Signed-off-by: Hariprasad Kelam <hkelam@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/52572
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/cgx.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/cgx.c b/drivers/net/ethernet/marvell/octeontx2/af/cgx.c
index 43d24ff20e72..5116b04a6dde 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/cgx.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/cgx.c
@@ -308,7 +308,7 @@ int cgx_lmac_addr_reset(u8 cgx_id, u8 lmac_id)
 	struct cgx *cgx_dev = cgx_get_pdata(cgx_id);
 	struct lmac *lmac = lmac_pdata(lmac_id, cgx_dev);
 	struct mac_ops *mac_ops;
-	u8 index = 0;
+	u8 index = 0, id;
 	u64 cfg;
 
 	if (!lmac)
@@ -320,7 +320,9 @@ int cgx_lmac_addr_reset(u8 cgx_id, u8 lmac_id)
 	 */
 	set_bit(0, lmac->mac_to_index_bmap.bmap);
 
-	index = lmac_id * lmac->mac_to_index_bmap.max + index;
+	id = get_sequence_id_of_lmac(cgx_dev, lmac_id);
+
+	index = id * lmac->mac_to_index_bmap.max + index;
 	cgx_write(cgx_dev, 0, (CGXX_CMRX_RX_DMAC_CAM0 + (index * 0x8)), 0);
 
 	/* Reset CGXX_CMRX_RX_DMAC_CTL0 register to default state */
-- 
2.31.1

